### PR TITLE
[LLVMExternalProjects] Drop the IREE LinalgTransformDialect

### PR DESCRIPTION
**Pull request by qedawkins**
The LinalgTransformDialect has become a maintenance burden without any users. This drops the dialect and cleans up some surrounding linalg ext code that has since been upstreamed.

 - Transform ops equivalent to lower_to_llvm now exist upstream
 - The linalg ext variant of the TransformDialectInterpreter pass has since been moved to IREE core

take_first, emit_remark, and the matcher infra + ops are still load bearing and transitioning to upstream is left as TODO.

### CODE REVIEW

**Review by stellaraccident — APPROVED** [review]
Strong +1 of no motivated users

**Comment by qedawkins** [comment]
Adding some people who I think have worked on this in the past to see if there is anything removed here that wants to be upstreamed first.

File: llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgTransform/CMakeLists.txt [hunk 1]
```diff
@@ -16,20 +6,6 @@ function(_add_transform_dialect_extension)
...
-  set(GEN_DOC_FILE ${IREE_DIALECTS_BINARY_DIR}/docs/Dialects/LinalgTransformOps.md)
-  add_custom_command(
-          OUTPUT ${GEN_DOC_FILE}
-          COMMAND ${CMAKE_COMMAND} -E copy
-                  ${CMAKE_CURRENT_BINARY_DIR}/LinalgTransformOps.md
-                  ${GEN_DOC_FILE}
-          DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/LinalgTransformOps.md)
-  add_custom_target(LinalgTransformOpsDocGen DEPENDS ${GEN_DOC_FILE})
-  add_dependencies(iree-dialects-doc LinalgTransformOpsDocGen)
-endfunction()
```
Comment (ScottTodd) [inline llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgTransform/CMakeLists.txt:31]: Can you also remove references to this dialect from the documentation generation script?

https://github.com/openxla/iree/blob/14ce232df01b8606a7f2c8820050008200128afc/docs/website/generate_extra_files.sh#L48-L62

Should just be a matter of removing the two lines that reference `LinalgTransformOps`.

File: llvm-external-projects/iree-dialects/BUILD.bazel
Comment (ScottTodd) [inline llvm-external-projects/iree-dialects/BUILD.bazel:1]: Seeing a few CI errors - want any help understanding / debugging those?

* Bazel: https://github.com/openxla/iree/actions/runs/6356500159/job/17266265320?pr=15067
* ASan: https://github.com/openxla/iree/actions/runs/6356500159/job/17266267038?pr=15067

**Review by ScottTodd — APPROVED** [review]
Cool - always happy to see code cleanup, if this is indeed no longer being used or actively developed.

File: llvm-external-projects/iree-dialects/BUILD.bazel
Comment (qedawkins) [reply llvm-external-projects/iree-dialects/BUILD.bazel:1]: Working through them little by little, but any help would be great!

File: llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgTransform/CMakeLists.txt (hunk 1 above)
Comment (qedawkins) [reply llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgTransform/CMakeLists.txt:31]: I missed those, thanks

**Comment by qedawkins** [comment]
Darn, I wanted to try and drop the interpreter pass in linalg-transform-ext, but it looks like it's needed to test some of the linalg-ext transform ops
```
 iree-dialects-opt: Unknown command line argument '--transform-dialect-interpreter'.  Try: '/work/build-asan/llvm-project/bin/iree-dialects-opt --help'
iree-dialects-opt: Did you mean '--transform-dialect-drop-schedule'?
FileCheck error: '<stdin>' is empty.
FileCheck command line:  FileCheck /work/llvm-external-projects/iree-dialects/test/Dialect/iree_linalg_ext/foreach-thread-to-async.mlir
```

**Comment by nicolasvasilache** [comment]
Thanks for the cleanup!

**Review by nicolasvasilache — APPROVED** [review]

**Comment by qedawkins** [comment]
I'll land this tomorrow if there are no more comments.

**Comment by ftynse** [comment]
As an FYI, I don't have a say in what IREE does:

- `emit_remark` is available in the "test transform" extension upstream (but downstreams normally don't load test extensions); this may be moved into the "main" transform dialect IMO.
- Matcher infra + ops should not be upstreamed; there is an arguably better alternative with matcher ops that do not require the awkward C++ interfacing. It is poorly documented, but there are tests https://github.com/llvm/llvm-project/tree/main/mlir/test/Integration/Dialect/Transform, and I committed enough matchers to replace all usage of the C++ matching as of ~May 2023. I didn't have the time to write a public design doc, but if there is interest in using this, I may be able to find the time or discuss with someone.
- There shouldn't be a need to have more than one transform interpreter pass in the project unless passes somehow change the interpreter logic. The base interpreter works with any ops implementing transform interfaces as long as the dialect extensions providing those ops correctly register dependent and generated dialects.

**Comment by qedawkins** [comment]
> As an FYI, I don't have a say in what IREE does:
> 
>     * `emit_remark` is available in the "test transform" extension upstream (but downstreams normally don't load test extensions); this may be moved into the "main" transform dialect IMO.
> 
>     * Matcher infra + ops should not be upstreamed; there is an arguably better alternative with matcher ops that do not require the awkward C++ interfacing. It is poorly documented, but there are tests https://github.com/llvm/llvm-project/tree/main/mlir/test/Integration/Dialect/Transform, and I committed enough matchers to replace all usage of the C++ matching as of ~May 2023. I didn't have the time to write a public design doc, but if there is interest in using this, I may be able to find the time or discuss with someone.
> 
>     * There shouldn't be a need to have more than one transform interpreter pass in the project unless passes somehow change the interpreter logic. The base interpreter works with any ops implementing transform interfaces as long as the dialect extensions providing those ops correctly register dependent and generated dialects.

Thanks for the notes. For the matcher infra, I didn't specifically mean moving it upstream, but instead transitioning to use what was added upstream (and if needed upstream any missing pieces). Definitely was a bit unclear looking back at the wording. There has been some light discussion about how to transition that C++, so there is definitely interest in a discussion at some point. I was thinking of re-evaluating how we handle that matching after #14788 and related PRs land.

Because llvm-external-projects is built independently of IREE (and used to test some linalg-ext ops) I assumed the custom interpreter pass similar to what we have in codegen is required for testing, but if the upstream one works that would be great! I can try that in a follow up.

> As an FYI, I don't have a say in what IREE does:

(also if the pings were annoying I am sorry about that. I just wanted to make sure that if there were anything left here still wanted upstream that it was captured).

**Comment by MaheshRavishankar** [comment]
> As an FYI, I don't have a say in what IREE does:
> 
You absolutely do have a say in what IREE does. Any feedback/viewpoint that helps lead IREE to a better state are welcome!

**Pull request merged**